cmake/IntegratedOpenCL.cmake

```diff
@@ -16,20 +16,28 @@ include(FetchContent)
 FetchContent_Declare(OpenCL-Headers GIT_REPOSITORY ${OPENCL_HEADER_REPOSITORY} GIT_TAG ${OPENCL_HEADER_TAG})
 FetchContent_GetProperties(OpenCL-Headers)
 if(NOT OpenCL-Headers_POPULATED)
-  FetchContent_Populate(OpenCL-Headers)
+  FetchContent_MakeAvailable(OpenCL-Headers)
   message(STATUS "Populated OpenCL Headers")
 endif()
 set(OPENCL_ICD_LOADER_HEADERS_DIR ${opencl-headers_SOURCE_DIR} CACHE PATH "") # for OpenCL ICD Loader
 set(OpenCL_INCLUDE_DIR ${opencl-headers_SOURCE_DIR} CACHE PATH "") # for Boost::Compute
 
-FetchContent_Declare(OpenCL-ICD-Loader GIT_REPOSITORY ${OPENCL_LOADER_REPOSITORY} GIT_TAG ${OPENCL_LOADER_TAG})
+if (CMAKE_VERSION VERSION_GREATER_EQUAL 3.28)
```

Comment:
Or we can bump minimum required version.

Reply:
I support just increasing the project-wide minimum CMake version (in both `CMakeLists.txt` and `pyproject.toml`) to 3.28.

I like this advice from https://cliutils.gitlab.io/modern-cmake/chapters/intro/installing.html.

> _Your CMake version should be newer than your compiler. It should be newer than the libraries you are using (especially Boost). **New versions work better for everyone.**_
> 
> _If you have a built in copy of CMake, it isn't special or customized for your system. You can easily install a new one instead, either on the system level or the user level. Feel free to instruct your users here if they complain about a CMake requirement being set too high._

It's true that 3.28.0 has only been out for 8 months (https://github.com/Kitware/CMake/releases/tag/v3.28.0), but it's already in wide use.

* for Mac users, Homebrew and macports already have v3.30.0
* for Windows users, Chocolatey has v3.30.0 ([link](https://community.chocolatey.org/packages/cmake)) and CMake official binaries have v3.30.0 ([link](https://cmake.org/download/))

The biggest disruption would be for Linux users. If you use `apt-get install cmake` on Ubuntu 22.04 today, you'll get v3.22.1. See the table at the bottom of https://cliutils.gitlab.io/modern-cmake/chapters/intro/installing.html or at https://repology.org/project/cmake/versions.

Linux users on distros with older versions would have to either install it a different way (like with `pip` or `conda`) or like this:

https://github.com/microsoft/LightGBM/blob/f8ec57b8eba69b7371324bba1d158f850802c782/docker/dockerfile-cli#L19-L22

I don't think that's too bad... especially since to build the Python package specifically, `scikit-build-core` will handle automatically `pip`-installing `cmake` if a new-enough version is not found.

And I don't think this would just be for the benefit of integrated OpenCL wheels... there are lots of other benefits we'd get by upgrading to a newer minimum version. See https://cliutils.gitlab.io/modern-cmake/chapters/intro/newcmake.html.

If you feel that this is too much of a risk, then I'd support changing the minimum in `CMakeLists.txt` conditionally, like

```cmake
if(__INTEGRATE_OPENCL)
  cmake_minimum_required(VERSION 3.18)
else()
  cmake_minimum_required(VERSION 3.28)
endif()
```

Reply:
Thanks a lot for your opinion and especially for the link to short version of CMake Changelog!

> `scikit-build-core` will handle automatically `pip`-installing `cmake` if a new-enough version is not found.

Cool feature of `scikit-build-core`! 👍 

I'll go ahead and bump CMake version. Due to the lack of working hands here, I think it's OK to bump version unconditionally.